File: source/go/string_permutations.go
package main

import "fmt"

func permutations(s string) []string {
	var results []string
	chars := []rune(s)

	var permute func(arr []rune, start int)
	permute = func(arr []rune, start int) {
		if start == len(arr)-1 {
			results = append(results, string(arr))
			return
		}
		for i := start; i < len(arr); i++ {
			arr[start], arr[i] = arr[i], arr[start]
			permute(arr, start+1)
			arr[start], arr[i] = arr[i], arr[start]
		}
	}
	permute(chars, 0)
	return results
}

func main() {
	perms := permutations("abc")
	for _, p := range perms {
		fmt.Println(p)
	}
}

File: source/rust/string_permutations.rs
fn permute(arr: &mut Vec<char>, start: usize, results: &mut Vec<String>) {
    if start == arr.len() - 1 {
        results.push(arr.iter().collect());
        return;
    }
    for i in start..arr.len() {
        arr.swap(start, i);
        permute(arr, start + 1, results);
        arr.swap(start, i);
    }
}

fn main() {
    let mut chars: Vec<char> = "abc".chars().collect();
    let mut results = Vec::new();
    permute(&mut chars, 0, &mut results);
    for p in &results {
        println!("{}", p);
    }
}
